File: android/app/src/main/java/com/steve1316/uma_android_automation/llm/GroundingVerifier.kt
package com.steve1316.uma_android_automation.llm

/**
 * Lightweight post-generation hallucination guard.
 *
 * Tokenizes the model's answer into content-word unigrams and checks what fraction of them appear verbatim in any
 * of the retrieved source chunks the prompt was grounded on. Below the threshold, the orchestrator discards the
 * generated answer and falls back to showing the top retrieved chunk verbatim — the well-studied failure mode for
 * sub-3B instruct models is confident confabulation, and n-gram overlap catches a meaningful fraction of it at
 * negligible cost.
 *
 * Not a replacement for stronger verification (NLI, citation-quote matching); intended as the cheap first layer.
 */
object GroundingVerifier {
    /** Default unigram-overlap threshold — hand-picked to allow natural paraphrase while rejecting invention. */
    const val DEFAULT_THRESHOLD = 0.4f

    /** English stopwords stripped before overlap computation so common function words don't carry the ratio. */
    private val STOPWORDS = setOf(
        "a", "an", "the", "is", "are", "was", "were", "be", "been", "being", "and", "or", "but", "if", "then",
        "to", "of", "in", "on", "for", "with", "by", "at", "as", "from", "into", "about", "it", "its", "this",
        "that", "these", "those", "you", "your", "yours", "we", "our", "ours", "they", "them", "their", "theirs",
        "he", "she", "him", "her", "his", "hers", "i", "me", "my", "mine", "not", "no", "yes", "do", "does",
        "did", "doing", "done", "have", "has", "had", "having", "will", "would", "should", "could", "can", "may",
        "might", "must", "shall", "so", "than", "too", "very", "just", "also", "such", "only", "some", "any",
    )

    /**
     * Check whether [answer] is sufficiently supported by the provided [context] chunks.
     *
     * @param answer Generated text from the LLM.
     * @param context Source chunks the prompt grounded the LLM on.
     * @param threshold Minimum fraction of answer content-unigrams that must be present in at least one context chunk.
     * @return true if overlap ≥ [threshold]; false otherwise.
     */
    fun isGrounded(answer: String, context: List<String>, threshold: Float = DEFAULT_THRESHOLD): Boolean {
        return overlap(answer, context) >= threshold
    }

    /**
     * Compute the unigram overlap ratio without applying the threshold. Useful for logging / tuning.
     *
     * @param answer Generated text from the LLM.
     * @param context Source chunks.
     * @return Ratio in [0.0, 1.0]. Returns 1.0 for an empty answer since "nothing" trivially matches.
     */
    fun overlap(answer: String, context: List<String>): Float {
        val answerTokens = contentTokens(answer)
        if (answerTokens.isEmpty()) return 1f
        val contextTokens = HashSet<String>()
        for (c in context) contextTokens.addAll(contentTokens(c))
        var matched = 0
        for (t in answerTokens) if (contextTokens.contains(t)) matched += 1
        return matched.toFloat() / answerTokens.size
    }

    private fun contentTokens(text: String): Set<String> {
        val out = HashSet<String>()
        val lower = text.lowercase()
        val sb = StringBuilder()
        for (ch in lower) {
            if (ch.isLetterOrDigit()) {
                sb.append(ch)
            } else {
                if (sb.isNotEmpty()) {
                    val token = sb.toString()
                    if (token.length > 1 && token !in STOPWORDS) out.add(token)
                    sb.clear()
                }
            }
        }
        if (sb.isNotEmpty()) {
            val token = sb.toString()
            if (token.length > 1 && token !in STOPWORDS) out.add(token)
        }
        return out
    }
}

File: src/lib/chat/groundingVerifier.ts
/**
 * TypeScript port of GroundingVerifier.kt.
 *
 * Lightweight post-generation hallucination guard: tokenizes the model's answer into content-word unigrams and
 * checks what fraction of them appear verbatim in any of the retrieved source chunks. Below the threshold, the
 * orchestrator discards the generated answer and falls back to showing the top retrieved chunk verbatim.
 *
 * Identical algorithm to the Kotlin original so behavior matches the existing 5 JUnit cases.
 */

/** Default unigram-overlap threshold — hand-picked to allow natural paraphrase while rejecting invention. */
export const DEFAULT_THRESHOLD = 0.4

/** Lower threshold the orchestrator actually uses, since summary-style answers paraphrase and reduce overlap. */
export const SUMMARY_THRESHOLD = 0.3

/** English stopwords stripped before overlap computation so common function words don't carry the ratio. */
const STOPWORDS = new Set<string>([
    "a", "an", "the", "is", "are", "was", "were", "be", "been", "being", "and", "or", "but", "if", "then",
    "to", "of", "in", "on", "for", "with", "by", "at", "as", "from", "into", "about", "it", "its", "this",
    "that", "these", "those", "you", "your", "yours", "we", "our", "ours", "they", "them", "their", "theirs",
    "he", "she", "him", "her", "his", "hers", "i", "me", "my", "mine", "not", "no", "yes", "do", "does",
    "did", "doing", "done", "have", "has", "had", "having", "will", "would", "should", "could", "can", "may",
    "might", "must", "shall", "so", "than", "too", "very", "just", "also", "such", "only", "some", "any",
])

/**
 * Compute the unigram overlap ratio between [answer] and the union of [context] chunks. Returns 1.0 for an empty
 * answer (nothing trivially matches) and a value in [0, 1] otherwise.
 */
export function overlap(answer: string, context: string[]): number {
    const answerTokens = contentTokens(answer)
    if (answerTokens.size === 0) return 1
    const contextTokens = new Set<string>()
    for (const c of context) for (const t of contentTokens(c)) contextTokens.add(t)
    let matched = 0
    for (const t of answerTokens) if (contextTokens.has(t)) matched += 1
    return matched / answerTokens.size
}

/** True when [overlap] meets or exceeds [threshold]. */
export function isGrounded(answer: string, context: string[], threshold: number = DEFAULT_THRESHOLD): boolean {
    return overlap(answer, context) >= threshold
}

function contentTokens(text: string): Set<string> {
    const out = new Set<string>()
    const lower = text.toLowerCase()
    let buf = ""
    for (const ch of lower) {
        if (isLetterOrDigit(ch)) {
            buf += ch
        } else {
            if (buf.length > 0) {
                if (buf.length > 1 && !STOPWORDS.has(buf)) out.add(buf)
                buf = ""
            }
        }
    }
    if (buf.length > 1 && !STOPWORDS.has(buf)) out.add(buf)
    return out
}

function isLetterOrDigit(ch: string): boolean {
    return /[\p{L}\p{N}]/u.test(ch)
}
